test/unit/index/test-rangesegment.cpp

```diff
@@ -0,0 +1,104 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/RAJA.hpp"
+#include "gtest/gtest.h"
+
+template<typename T>
+class RangeSegmentTest : public ::testing::Test {};
+
+using MyTypes = ::testing::Types<RAJA::Index_type,
+                                 char, 
+                                 unsigned char,
+				 short,
+				 unsigned short,
+				 int, 
+				 unsigned int,
+				 long,
+				 unsigned long,
+				 long int,
+				 unsigned long int,
+				 long long,
+				 unsigned long long>;
+
+TYPED_TEST_CASE(RangeSegmentTest, MyTypes);
+
+TYPED_TEST(RangeSegmentTest, Constructors)
+{
+  RAJA::TypedRangeSegment<TypeParam> first(0, 10);
+  RAJA::TypedRangeSegment<TypeParam> copied(first);
+
+  ASSERT_EQ(first, copied);
+
+  RAJA::TypedRangeSegment<TypeParam> moved(std::move(first));
+
+  ASSERT_EQ(moved, copied);
+
+  if(std::is_signed<TypeParam>::value){
+    RAJA::TypedRangeSegment<TypeParam> r1(-10, 7);
+    RAJA::TypedRangeSegment<TypeParam> r2(0, -50);
+    ASSERT_EQ(17, r1.size());
+    ASSERT_EQ(-50, r2.size());
```

Comment:
@rhornung67 @mdavis36  - is this correct? I thought that the size of segment would be 0 when end < begin?

Reply:
I thought so too, but that's not the case. Not sure how we can change this since size is computed by subtracting iterators. @mdavis36 any thoughts?

Reply:
@davidbeckingsale @rhornung67 This seems to be related to an issue previously discussed in this PR. Since RangeSegments are used for positive iteration, it would be ideal to ensure that users aren't defining RangeSegments with begin > end. I can add a static_assert and an exception check in the constructor so we can catch it at compile or runtime.

Reply:
I like that idea! Go for it.